### PR TITLE
chore: migrate to singularity --nvccli [DET-9081]

### PR DESCRIPTION
**Pull request by phillip-gaisford**
## Description

To quote the ticket: migrate to singularity run calls from `--nvccli` from `--nv` as it more closely mirrors docker's GPU support

In order to avoid this error:

image

I am also specifying the qualifier `--userns`; this introduces some restrictions, but doesn't seem to cause a problem with the DAI environment images used so far.

## Test Plan

Tested on casablanca-login using:

`apptainer version 1.2.0~rc.1-1`

Performed some one and two-GPU experiments. By injecting the command nvidia-smi into the wrapper scripts, I was able to verify that only the expected GPUs were visible.

Start the agent:

`srun -N1 --gpus-per-node=4 /mnt/lustre/foundation_engineering/gaisford/determined-agent --master-host=casablanca-login  --master-port=8086 --resource-pool=default --container-runtime=singularity --slot-type=cuda --debug=false`

This provides four GPU slots.
Start the DAI master:

./tools/slurmcluster.sh -A casablanca-login

Configure an experiment to request two GPUs, and run two instances concurrently:

```
name: cifar10_pytorch_distributed
hyperparameters:
  learning_rate: 1.0e-4
  learning_rate_decay: 1.0e-6
  layer1_dropout: 0.25
  layer2_dropout: 0.25
  layer3_dropout: 0.5
  global_batch_size: 512 # Per-GPU batch size of 32
resources:
  slots_per_trial: 2
records_per_epoch: 50000
searcher:
  name: single
  metric: validation_error
  max_length:
    epochs: 32
entrypoint: model_def:CIFARTrial
min_validation_period:
  epochs: 1
```

From the log files, see that in each instance only two GPUs are accessible, different in each experiment instance:

Instance one:

```
[2023-07-07T19:54:57.066322Z] 791da9e4 || DEBUG: singularity run \
	--writable-tmpfs \
	--pwd /run/determined/workdir \
	--env-file /tmp/determined-gaisford-node002/392168593-791da9e4-732e-4605-8dff-a531b97328a3/envfile \
	--bind /tmp/determined-gaisford-node002/392168593-791da9e4-732e-4605-8dff-a531b97328a3/archives/etc/ssh/ssh_config:/etc/ssh/ssh_config \
	--bind /tmp/determined-gaisford-node002/392168593-791da9e4-732e-4605-8dff-a531b97328a3/archives/opt/determined:/opt/determined \
	--bind /tmp/determined-gaisford-node002/392168593-791da9e4-732e-4605-8dff-a531b97328a3/archives/run/determined:/run/determined \
	--bind /etc/hosts:/etc/hosts \
	--bind /mnt/lustre/foundation_engineering/determined-cp:/determined_shared_fs \
	--nv \
	--nvccli \
	--userns \
	--contain docker://determinedai/environments:cuda-11.3-pytorch-1.12-tf-2.8-gpu-9d07809 /run/determined/singularity-entrypoint-wrapper.sh /run/determined/train/entrypoint.sh
[2023-07-07T19:54:56.966625Z] 791da9e4 || INFO: Resources for Trial 2063 (Experiment 366) have started
[2023-07-07T19:54:58.791513Z] 791da9e4 || INFO: Using cached SIF image
[2023-07-07T19:54:58.792863Z] 791da9e4 || INFO: WARNING: Overriding HOME environment variable with APPTAINERENV_HOME is not permitted
[2023-07-07T19:54:58.849884Z] 791da9e4 || INFO: fuse: warning: library too old, some operations may not work
[2023-07-07T19:55:01.879603Z] 791da9e4 || INFO: ==========
[2023-07-07T19:55:01.879618Z] 791da9e4 || INFO: == CUDA ==
[2023-07-07T19:55:01.879701Z] 791da9e4 || INFO: ==========
[2023-07-07T19:55:01.886888Z] 791da9e4 || INFO: CUDA Version 11.3.1
[2023-07-07T19:55:01.889778Z] 791da9e4 || INFO: Container image Copyright (c) 2016-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
[2023-07-07T19:55:01.891478Z] 791da9e4 || INFO: This container image and its contents are governed by the NVIDIA Deep Learning Container License.
[2023-07-07T19:55:01.891489Z] 791da9e4 || INFO: By pulling and using the container, you accept the terms and conditions of this license:
[2023-07-07T19:55:01.891498Z] 791da9e4 || INFO: https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license
[2023-07-07T19:55:01.891506Z] 791da9e4 || INFO: A copy of this license is made available in this container at /NGC-DL-CONTAINER-LICENSE for your convenience.
[2023-07-07T19:55:01.946312Z] 791da9e4 || INFO:   ROCR_VISIBLE_DEVICES: 
[2023-07-07T19:55:01.946314Z] 791da9e4 || INFO:   CUDA_VISIBLE_DEVICES: 
[2023-07-07T19:55:01.946317Z] 791da9e4 || INFO: NVIDIA_VISIBLE_DEVICES: all
[2023-07-07T19:55:01.946321Z] 791da9e4 || INFO:           DET_SLOT_IDS: [0,1]
[2023-07-07T19:55:03.412870Z] 791da9e4 || INFO: GPU 0: Tesla V100-SXM2-16GB (UUID: GPU-ee310a06-c7fb-f921-f137-d605913909b1)
[2023-07-07T19:55:03.412903Z] 791da9e4 || INFO: GPU 1: Tesla V100-SXM2-16GB (UUID: GPU-b4b1244f-a384-ac0a-2933-ee2f2e10ae26)

```

Instance two:

```
[2023-07-07T19:56:47.264986Z] 413dfa8e || DEBUG: singularity run \
	--writable-tmpfs \
	--pwd /run/determined/workdir \
	--env-file /tmp/determined-gaisford-node002/218428094-413dfa8e-2837-4b3b-a9e3-24831cd579a7/envfile \
	--bind /tmp/determined-gaisford-node002/218428094-413dfa8e-2837-4b3b-a9e3-24831cd579a7/archives/etc/ssh/ssh_config:/etc/ssh/ssh_config \
	--bind /tmp/determined-gaisford-node002/218428094-413dfa8e-2837-4b3b-a9e3-24831cd579a7/archives/opt/determined:/opt/determined \
	--bind /tmp/determined-gaisford-node002/218428094-413dfa8e-2837-4b3b-a9e3-24831cd579a7/archives/run/determined:/run/determined \
	--bind /etc/hosts:/etc/hosts \
	--bind /mnt/lustre/foundation_engineering/determined-cp:/determined_shared_fs \
	--nv \
	--nvccli \
	--userns \
	--contain docker://determinedai/environments:cuda-11.3-pytorch-1.12-tf-2.8-gpu-9d07809 /run/determined/singularity-entrypoint-wrapper.sh /run/determined/train/entrypoint.sh
[2023-07-07T19:56:47.279204Z] 413dfa8e || INFO: Resources for Trial 2064 (Experiment 367) have started
[2023-07-07T19:56:48.920342Z] 413dfa8e || INFO: Using cached SIF image
[2023-07-07T19:56:48.921622Z] 413dfa8e || INFO: WARNING: Overriding HOME environment variable with APPTAINERENV_HOME is not permitted
[2023-07-07T19:56:48.978167Z] 413dfa8e || INFO: fuse: warning: library too old, some operations may not work
[2023-07-07T19:56:51.438233Z] 413dfa8e || INFO: ==========
[2023-07-07T19:56:51.438248Z] 413dfa8e || INFO: == CUDA ==
[2023-07-07T19:56:51.438325Z] 413dfa8e || INFO: ==========
[2023-07-07T19:56:51.445390Z] 413dfa8e || INFO: CUDA Version 11.3.1
[2023-07-07T19:56:51.448219Z] 413dfa8e || INFO: Container image Copyright (c) 2016-2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
[2023-07-07T19:56:51.449831Z] 413dfa8e || INFO: This container image and its contents are governed by the NVIDIA Deep Learning Container License.
[2023-07-07T19:56:51.449843Z] 413dfa8e || INFO: By pulling and using the container, you accept the terms and conditions of this license:
[2023-07-07T19:56:51.449849Z] 413dfa8e || INFO: https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license
[2023-07-07T19:56:51.449857Z] 413dfa8e || INFO: A copy of this license is made available in this container at /NGC-DL-CONTAINER-LICENSE for your convenience.
[2023-07-07T19:56:51.502424Z] 413dfa8e || INFO:   ROCR_VISIBLE_DEVICES: 
[2023-07-07T19:56:51.502427Z] 413dfa8e || INFO:   CUDA_VISIBLE_DEVICES: 
[2023-07-07T19:56:51.502431Z] 413dfa8e || INFO: NVIDIA_VISIBLE_DEVICES: all
[2023-07-07T19:56:51.502493Z] 413dfa8e || INFO:           DET_SLOT_IDS: [2,3]
[2023-07-07T19:56:52.981627Z] 413dfa8e || INFO: GPU 0: Tesla V100-SXM2-16GB (UUID: GPU-8536a49c-3b4c-c98c-7f14-b0a7c7755d64)
[2023-07-07T19:56:52.981643Z] 413dfa8e || INFO: GPU 1: Tesla V100-SXM2-16GB (UUID: GPU-df5fc6c8-540a-ed75-f907-80647319fb3b)

```

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
https://hpe-aiatscale.atlassian.net/browse/DET-9081

### CODE REVIEW

File: agent/pkg/singularity/singularity.go [hunk 1]
```diff
@@ -310,9 +310,9 @@ func (s *SingularityClient) RunContainer(
 		}
 	}
 	if len(cudaVisibleDevices) > 0 {
-		// TODO(DET-9081): We need to move to --nvccli --nv, because --nv does not provide
-		// sufficient isolation (e.g., nvidia-smi see all GPUs on the machine, not just ours).
-		args = append(args, "--nv")
+		// Using --userns to avoid this error when using --nvccli:
+		// FATAL:   nvidia-container-cli not allowed in setuid mode
+		args = append(args, "--nv", "--nvccli", "--userns", "--contain")
```
Comment (phillip-gaisford) [inline agent/pkg/singularity/singularity.go:315]:
```
  -c, --contain                       use minimal /dev and empty other
                                      directories (e.g. /tmp and $HOME)
                                      instead of sharing filesystems from
                                      your host

```
Is this going to work for us?

**Comment by stoksc** [comment]
@phillip-gaisford you probably want eyes on this that aren't me, not a singularity expert.

**Comment by stoksc** [comment]
that is to say, lgtm but what do i know

**Comment by rcorujo** [comment]
Can you add the ticket number to the end of the PR title in square brackets? 

Ticket DET-9081's title says ```migrate to singularity run calls from `--nvccli` from `--nv` as it more closely mirrors docker's GPU support```, which doesn't provide a whole lot of background information.

Your description says, ```Migrate singularity run calls to use --nvccli, for better isolation.```.

Can you elaborate what ```better isolation``` means?  What is the problem that this ticket is addressing?  Is it that one job can use a partial set of GPUs on a node and another job is able to use the remaining set of GPUs on that node?  Was that not the case before?



Isn't the "DEBUG" redundant now that you've included DET_SLOT_IDS in the "INFO"?

```
[2023-07-07T19:55:01.946303Z] 791da9e4 || INFO: DEBUG: Will utilize slots DET_SLOT_IDS [0,1]
[2023-07-07T19:55:01.946312Z] 791da9e4 || INFO:   ROCR_VISIBLE_DEVICES: 
[2023-07-07T19:55:01.946314Z] 791da9e4 || INFO:   CUDA_VISIBLE_DEVICES: 
[2023-07-07T19:55:01.946317Z] 791da9e4 || INFO: NVIDIA_VISIBLE_DEVICES: all
[2023-07-07T19:55:01.946321Z] 791da9e4 || INFO:           DET_SLOT_IDS: [0,1]
```

**Comment by phillip-gaisford** [comment]
> Can you add the ticket number to the end of the PR title in square brackets?

Done (also the ticket ID is in the branch name)

**Comment by phillip-gaisford** [comment]
> Can you elaborate what `better isolation` means? What is the problem that this ticket is addressing? Is it that one job can use a partial set of GPUs on a node and another job is able to use the remaining set of GPUs on that node? Was that not the case before?

That was my take on the description from the ticket, a facet I have now removed.

A feature of the current implementation, and my guess as to the reason for the ticket, is that although CUDA_VISIBLE_DEVICES is correctly set in the container, any other GPUs in the allocation are still visible in the container, which presumably is undesirable. Use of `--contain` does away with this, at a certain cost.

**Comment by phillip-gaisford** [comment]
> Isn't the "DEBUG" redundant now that you've included DET_SLOT_IDS in the "INFO"?

Correct. This output came from a modified version of the wrapper script -- you won't see this in the product. FYI, the master sends a new copy of the wrapper with each experiment; it is quite convenient to add debug lines to it.

**Comment by rcorujo** [comment]
> > Can you elaborate what `better isolation` means? What is the problem that this ticket is addressing? Is it that one job can use a partial set of GPUs on a node and another job is able to use the remaining set of GPUs on that node? Was that not the case before?
> 
> That was my take on the description from the ticket, a facet I have now removed.
> 
> A feature of the current implementation, and my guess as to the reason for the ticket, is that although CUDA_VISIBLE_DEVICES is correctly set in the container, any other GPUs in the allocation are still visible in the container, which presumably is undesirable. Use of `--contain` does away with this, at a certain cost.

Did you mean to say, "any other GPUs ```that are NOT part of``` the allocation are still visible in the container"?  Certainly you want GPUs that are part of the allocation to be visible in the container, but not those that you did not allocate, right?

**Comment by rcorujo** [comment]
If I'm understanding your testing section correctly, you ran two instances of a job on the same node. Did you test a distributed job that runs on two different nodes?

**Comment by phillip-gaisford** [comment]
> Did you mean to say, "any other GPUs `that are NOT part of` the allocation are still visible in the container"

No. By allocation I am referring to the set of GPUs allocated to the job hosting the agent (the `srun` in the given example). Experiments will use a subset of those GPUs, and ideally (I suppose) not have any access to the remainder.

**Review by rcorujo — APPROVED** [review]
